rust/mdb_shard/src/merging.rs

```diff
@@ -33,134 +33,129 @@ fn write_shard(target_directory: &Path, data: &[u8]) -> Result<(MerkleHash, Path
-                    // There should always be something in the queue
-                    let alt_si = queue.pop().unwrap();
-
-                    alt_data.clear();
-                    std::fs::File::open(&alt_si.path)?.read_to_end(&mut alt_data)?;
-                    shards_to_remove.push(alt_si.path);
-
-                    out_data.clear();
-
-                    // Merge these in to the current shard.
-                    let new_shard = shard_set_union(
-                        &cur_si.shard,
-                        &mut Cursor::new(&cur_data),
-                        &alt_si.shard,
-                        &mut Cursor::new(&alt_data),
-                        &mut out_data,
-                    )?;
-
-                    if new_shard.num_bytes() > target_min_size || queue.is_empty() {
-                        let (shard_hash, full_file_name) =
-                            write_shard(result_directory, &out_data)?;
-
-                        let si = MDBShardFile {
-                            shard_hash,
-                            path: std::fs::canonicalize(&full_file_name)?,
-                            shard: new_shard,
-                        };
-                        info!(
-                            "Created merged shard {:?} from shards {:?}",
-                            &si.path, &shards_to_remove
-                        );
-
-                        completed_shards.push(si);
-
-                        // Delete all the previous ones, now that this has been successfully written out.
-                        for path in shards_to_remove {
-                            info!("Removing shard {:?}", &path);
-                            std::fs::remove_file(&path)?;
-                        }
-
-                        break;
-                    } else {
-                        swap(&mut cur_data, &mut out_data);
-                        cur_si.shard = new_shard;
-                        cur_si.path = PathBuf::default();
-                        cur_si.shard_hash = MerkleHash::default();
-                    }
-                }
+            let (shard_hash, full_file_name) = write_shard(session_directory, &cur_data)?;
+
+            let new_sfi = MDBShardFile {
+                shard_hash,
+                path: std::fs::canonicalize(&full_file_name)?,
+                shard: cur_shard_info,
+            };
+            debug!(
+                "Created merged shard {:?} from shards {:?}",
+                &new_sfi.path, &files_to_delete
+            );
+
+            // Put the new shard on the return stack and move to the next unmerged one
+            finished_shards.push(new_sfi);
+        }
+
+        // Delete the old ones.
+        for p in files_to_delete.iter() {
+            // In rare cases, there could be empty shards or shards with
+            // duplicate entries and we don't want to delete any shards
+            // we've already finished
+            if *p != finished_shards.last().unwrap().path {
```

Comment:
Such rare case can happen in any order and may not be the last finished shard?

Reply:
True...   It would mean that the previous shard information would have been merged into the current one, so it would be okay to delete in theory, but this would mean the list returned would be invalid.  I'll change it to check all of them.

Reply:
Fixed.